<launch>
  <arg name="UAV_NAME" default="$(optenv UAV_NAME)"/>

  <arg name="standalone" default="false"/>

  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg UAV_NAME)_uvdar_nodelet_manager"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>

  <arg name="threshold" default="50"/>

  <arg name="left" default="$(optenv BLUEFOX_UV_LEFT)"/> <arg name="right" default="$(optenv BLUEFOX_UV_RIGHT)"/>
  <arg name="left_camera_name" default="mv_$(arg left)"/>
  <arg name="right_camera_name" default="mv_$(arg right)"/>

  <arg name="calibrations_folder" default="$(find uvdar_gazebo_plugin)/test/detector/masks"/>


  <!-- Node Settings -->
  <arg name="output" default="screen"/>
  <arg name="debug" default="false"/>

  <arg name="publish" default="true"/>

  <arg name="sequence_file" default="$(find uvdar_core)/config/selected.txt"/>
  <arg name="model_file" default="$(find uvdar_core)/config/models/quadrotor_foursided_px4_motor_aligned.txt"/>

  <arg name="pub_tracking_stats" default="true"/>
  <arg name="draw_predict_window_sec" default="0.3"/>

    
  <arg name="max_px_shift_y" default="3"/>
  <arg name="max_px_shift_x" default="3"/>
  <arg name="max_zeros_consecutive" default="2"/>
  <arg name="max_ones_consecutive" default="2"/>
  <arg name="stored_seq_len_factor" default="1"/>
  <arg name="max_buffer_length" default="5000"/>
  <arg name="poly_order" default="2"/>
  <arg name="decay_factor" default="0.1"/>
  <arg name="confidence_probability" default="95.0"/>
  <arg name="allowed_BER_per_seq" default="0"/>

  <arg name="id1" value="0"/>
  <arg name="id2" value="1"/>
  <arg name="id3" value="2"/>
  <arg name="id4" value="3"/>
  <arg name="id5" value="4"/>
  <arg name="id6" value="5"/>
  <arg name="id7" value="6"/>
  <arg name="id8" value="7"/>
  <arg name="id9" value="8"/>
  <arg name="id10" value="9"/>
  <arg name="id11" value="10"/>
  <arg name="id12" value="11"/>
    
  <arg name="blink_throttle_rate" default="10"/>
    
  <group ns="$(arg UAV_NAME)">


      <!-- <node pkg="nodelet" type="nodelet" name="$(arg nodelet_manager)" args="manager" output="screen"  launch-prefix="xvfb-run gdb -batch -ex run -ex bt -/-args"> -->
    <node pkg="nodelet" type="nodelet" name="$(arg UAV_NAME)_uvdar_nodelet_manager" args="manager" output="screen"  launch-prefix="xvfb-run">
      <param name="num_worker_threads" value="8" />
    </node>

    <node name="bluefox_emulator" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBluefoxEmulator $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="bash -c 'sleep 0; $0 $@'">
      <rosparam param="camera_output_topics"> ["camera_left", "camera_right", "camera_back"] </rosparam>
      <rosparam param="calib_files"> ["default", "default", "default"] </rosparam>
      <remap from="~camera_left" to="/$(arg UAV_NAME)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right" to="/$(arg UAV_NAME)/uvdar_bluefox_right/image_raw"/>
      <remap from="~camera_back" to="/$(arg UAV_NAME)/uvdar_bluefox_back/image_raw"/>
      <remap from="~camera_left_transfer" to="/gazebo/$(arg UAV_NAME)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right_transfer" to="/gazebo/$(arg UAV_NAME)/uvdar_bluefox_right/image_raw"/>
      <remap from="~camera_back_transfer" to="/gazebo/$(arg UAV_NAME)/uvdar_bluefox_back/image_raw"/>
    </node>



    <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="xvfb-run"> --> 
    <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="false"  launch-prefix="bash -c 'sleep 0; $0 $@'"> 
    <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="debug_roslaunch"> --> 
      <!-- <node name="uv_detect" pkg="uvdar" type="uv_detector_node" output="screen" launch-prefix="urxvt -e gdb -q -x /home/viktor/gdb.cmds -/-args"> -->
        <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="valgrind -/-tool=callgrind -/-callgrind-out-file=/home/viktor/callgrind.out -/-instr-atstart=no -/-collect-atstart=yes"> --> 

      <param name="uav_name"		type="string" value="$(arg UAV_NAME)"/> 
      <param name="debug"		type="bool" value="$(arg debug)"/> 
      <param name="gui"             type="bool" value="false"/>
      <!-- <param name="publish_visualization"             type="bool" value="true"/> -->
      <param name="publish_visualization"             type="bool" value="false"/>
      <param name="publish"         type="bool" value="$(arg publish)"/>
      <param name="justReport" type="bool" value="true"/>
      <param name="threshold" type="int" value="$(arg threshold)"/>

      <param name="initial_delay" type="int" value="10"/>
      
      <rosparam param="camera_topics"> ["camera_left", "camera_right", "camera_back"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_left", "points_seen_right", "points_seen_back"] </rosparam>

      <!-- <param name="use_masks"         type="bool" value="false"/> -->
      <param name="use_masks"         type="bool" value="true"/>
      <rosparam param="mask_file_names" subst_value="True"> ["$(arg calibrations_folder)/blank_bluefox.png", "$(arg calibrations_folder)/blank_bluefox.png", "$(arg calibrations_folder)/blank_bluefox.png"] </rosparam>

      <param name="publish_sun_points"         type="bool" value="true"/>

      <remap from="~camera_left" to="/$(arg UAV_NAME)/uvdar_bluefox_left/image_raw"/>
      <remap from="~camera_right" to="/$(arg UAV_NAME)/uvdar_bluefox_right/image_raw"/>
      <remap from="~camera_back" to="/$(arg UAV_NAME)/uvdar_bluefox_back/image_raw"/>
      <remap from="~points_seen_left" to="/$(arg UAV_NAME)/uvdar/points_seen_left"/>
      <remap from="~points_seen_right" to="/$(arg UAV_NAME)/uvdar/points_seen_right"/>
      <remap from="~points_seen_back" to="/$(arg UAV_NAME)/uvdar/points_seen_back"/>
      <remap from="~points_seen_left/sun" to="/$(arg UAV_NAME)/uvdar/points_seen_left/sun"/>
      <remap from="~points_seen_right/sun" to="/$(arg UAV_NAME)/uvdar/points_seen_right/sun"/>
      <remap from="~points_seen_back/sun" to="/$(arg UAV_NAME)/uvdar/points_seen_back/sun"/>

      <remap from="~odometry" to="/$(arg UAV_NAME)/mrs_odometry/new_odom"/>
      <remap from="~imu" to="mavros/imu/data"/>
    </node>

    <node name="blink_processor" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="bash -c 'sleep 0; $0 $@'">

      <param name="uav_name" type = "string" value="$(arg UAV_NAME)"/>
      <param name="debug"		type="bool" value="false"/> 
      <param name="gui"             type="bool" value="false"/>
      <!-- <param name="gui"             type="bool" value="true"/> -->
      
      <!-- Changeable for each nodelet independently. - no arguments! -->
      <!-- <param name="publish_visualization"             type="bool" value="false"/> -->
      <param name="publish_visualization"             type="bool" value="false"/>
      <param name="use_camera_for_visualization"         type="bool" value="false"/>
      <param name="visualization_rate"         type="double" value="2"/>

      <param name="pub_tracking_stats"      type="bool" value="$(arg pub_tracking_stats)"/>
      <param name="draw_predict_window_sec"  type="double" value="$(arg draw_predict_window_sec)"/>
      <param name="use_4DHT"             type="bool" value="false"/>
    

      <param name="max_px_shift_x"          type="int"    value="$(arg max_px_shift_x)"/>
      <param name="max_px_shift_y"          type="int"    value="$(arg max_px_shift_y)"/>
      <param name="max_zeros_consecutive"   type="int"    value="$(arg max_zeros_consecutive)"/>
      <param name="max_ones_consecutive"    type="int"    value="$(arg max_ones_consecutive)"/>
      <param name="stored_seq_len_factor"   type="int"    value="$(arg stored_seq_len_factor)"/>  
      <param name="max_buffer_length"       type="int"    value="$(arg max_buffer_length)"/>  
      <param name="poly_order"              type="int"    value="$(arg poly_order)"/>
      <param name="decay_factor"            type="double" value="$(arg decay_factor)"/>    
      <param name="confidence_probability"  type="double" value="$(arg confidence_probability)"/>
      <param name="allowed_BER_per_seq"     type="int"    value="$(arg allowed_BER_per_seq)"/>
      
      <param name="sequence_file"         type="string" value="$(arg sequence_file)"/>

      <rosparam param="camera_topics"> ["camera_left", "camera_right", "camera_back"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_left", "points_seen_right", "points_seen_back"] </rosparam>
      <rosparam param="blinkers_seen_topics"> ["blinkers_seen_left", "blinkers_seen_right", "blinkers_seen_back"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimated_framerate_left", "estimated_framerate_right", "estimated_framerate_back"] </rosparam>
      <rosparam param="ami_logging_topics"> ["ami_logging_left", "ami_logging_right", "ami_logging_back"] </rosparam>
      <rosparam param="ami_all_seq_info_topics"> ["ami_all_seq_info_left", "ami_all_seq_info_right", "ami_all_seq_info_back"] </rosparam>
      <remap from="~camera_left" to="/$(arg UAV_NAME)/uvdar_bluefox_left/image_raw"/> -
      <remap from="~camera_right" to="/$(arg UAV_NAME)/uvdar_bluefox_right/image_raw"/>
      <remap from="~camera_back" to="/$(arg UAV_NAME)/uvdar_bluefox_back/image_raw"/>
      <remap from="~points_seen_left" to="/$(arg UAV_NAME)/uvdar/points_seen_left"/>
      <remap from="~points_seen_right" to="/$(arg UAV_NAME)/uvdar/points_seen_right"/>
      <remap from="~points_seen_back" to="/$(arg UAV_NAME)/uvdar/points_seen_back"/>
      <remap from="~blinkers_seen_left" to="/$(arg UAV_NAME)/uvdar/blinkers_seen_left"/>
      <remap from="~blinkers_seen_right" to="/$(arg UAV_NAME)/uvdar/blinkers_seen_right"/>
      <remap from="~blinkers_seen_back" to="/$(arg UAV_NAME)/uvdar/blinkers_seen_back"/>
      <remap from="~estimated_framerate_left" to="/$(arg UAV_NAME)/uvdar/estimated_framerate_left"/>
      <remap from="~estimated_framerate_right" to="/$(arg UAV_NAME)/uvdar/estimated_framerate_right"/>
      <remap from="~estimated_framerate_back" to="/$(arg UAV_NAME)/uvdar/estimated_framerate_back"/>
      <remap from="~ami_logging_left" to="/$(arg UAV_NAME)/uvdar/ami_logging_left"/>
      <remap from="~ami_logging_right" to="/$(arg UAV_NAME)/uvdar/ami_logging_right"/>
      <remap from="~ami_logging_back" to="/$(arg UAV_NAME)/uvdar/ami_logging_back"/>
      <remap from="~ami_all_seq_info_left" to="/$(arg UAV_NAME)/uvdar/ami_all_seq_info_left"/>
      <remap from="~ami_all_seq_info_right" to="/$(arg UAV_NAME)/uvdar/ami_all_seq_info_right"/>
      <remap from="~ami_all_seq_info_back" to="/$(arg UAV_NAME)/uvdar/ami_all_seq_info_back"/>
      <remap from="~visualization" to="/$(arg UAV_NAME)/uvdar/blink_visualization/image_raw"/>

    </node>

    <node
      name="uvcam_left_tf_$(arg UAV_NAME)"
      pkg="tf2_ros"
      type="static_transform_publisher"
      args="0.03  0.10 0.06 -0.3490658504 0.0 -1.57079632679 $(arg UAV_NAME)/fcu $(arg UAV_NAME)/uvcam_left"/>

    <node
      name="uvcam_right_tf_$(arg UAV_NAME)"
      pkg="tf2_ros"
      type="static_transform_publisher"
      args="0.03 -0.10 0.06 -2.792526803 0.0 -1.57079632679 $(arg UAV_NAME)/fcu $(arg UAV_NAME)/uvcam_right"/>

    <node
      name="uvcam_back_tf_$(arg UAV_NAME)"
      pkg="tf2_ros"
      type="static_transform_publisher"
      args="-0.10 0.0 -0.02 0.0 1.57079632679 3.14159265 $(arg UAV_NAME)/fcu $(arg UAV_NAME)/uvcam_back"/>
      
    <node name="throttle_blinkers_left" type="throttle" pkg="topic_tools"
      args="messages /$(arg UAV_NAME)/uvdar/blinkers_seen_left $(arg blink_throttle_rate) /$(arg UAV_NAME)/uvdar/blinkers_seen_left/throttled" />
    <node name="throttle_blinkers_right" type="throttle" pkg="topic_tools"
      args="messages /$(arg UAV_NAME)/uvdar/blinkers_seen_right $(arg blink_throttle_rate) /$(arg UAV_NAME)/uvdar/blinkers_seen_right/throttled" />
    <node name="throttle_blinkers_back" type="throttle" pkg="topic_tools"
      args="messages /$(arg UAV_NAME)/uvdar/blinkers_seen_back $(arg blink_throttle_rate) /$(arg UAV_NAME)/uvdar/blinkers_seen_back/throttled" />

    <node name="uvdar_pose_calculator_node" pkg="uvdar_core" type="uvdar_pose_calculator_node" output="screen" > 
      <param name="uav_name" type = "string" value="$(arg UAV_NAME)"/>
      <param name="debug"		type="bool" value="false"/> 
      <param name="gui"             type="bool" value="false"/>
      <param name="publish_visualization"             type="bool" value="false"/>
      <param name="publish_constituents"             type="bool" value="true"/>
      <rosparam param="signal_ids" subst_value="true"> [$(arg id1), $(arg id2), $(arg id3), $(arg id4), $(arg id5), $(arg id6), $(arg id7), $(arg id8), $(arg id9), $(arg id10), $(arg id11), $(arg id12)] </rosparam>


      <param name="profiling" type="bool" value="false"/>

      <param name="quadrotor" type="bool" value="false"/>
      <param name="custom_model" type="bool" value="true"/>
      <param name="model_file" type="string" value="$(arg model_file)"/>
      <param name="beacon" type="bool" value="false"/>

      <rosparam param="blinkers_seen_topics"> ["blinkers_seen_left", "blinkers_seen_right", "blinkers_seen_back"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimated_framerate_left", "estimated_framerate_right", "estimated_framerate_back"] </rosparam>
      <rosparam param="camera_frames" subst_value="true"> ["$(arg UAV_NAME)/uvcam_left", "$(arg UAV_NAME)/uvcam_right", "$(arg UAV_NAME)/uvcam_back"] </rosparam>

      <remap from="~blinkers_seen_left" to="/$(arg UAV_NAME)/uvdar/blinkers_seen_left/throttled"/>
      <remap from="~blinkers_seen_right" to="/$(arg UAV_NAME)/uvdar/blinkers_seen_right/throttled"/>
      <remap from="~blinkers_seen_back" to="/$(arg UAV_NAME)/uvdar/blinkers_seen_back/throttled"/>
      <remap from="~estimated_framerate_left" to="/$(arg UAV_NAME)/uvdar/estimated_framerate_left"/>
      <remap from="~estimated_framerate_right" to="/$(arg UAV_NAME)/uvdar/estimated_framerate_right"/>
      <remap from="~estimated_framerate_back" to="/$(arg UAV_NAME)/uvdar/estimated_framerate_back"/>

      <rosparam param="calib_files" subst_value="True"> ["default", "default", "default"] </rosparam>

      <remap from="~constituentHypotheses" to="/$(arg UAV_NAME)/uvdar/constituentHypotheses"/>
      <remap from="~measuredPoses" to="/$(arg UAV_NAME)/uvdar/measuredPoses"/>

      <param name="output_frame"      value="fcu"/>
        
    </node>

  </group>


</launch>
